kv/kvserver: TestLeaseQueueLeasePreferencePurgatoryError failed

kv/kvserver.TestLeaseQueueLeasePreferencePurgatoryError [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17644279?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17644279?buildTab=artifacts#/) on release-24.2 @ [07859fdcf04928e89e45720be32d70d3680de120](https://github.com/cockroachdb/cockroach/commits/07859fdcf04928e89e45720be32d70d3680de120):


```
=== RUN   TestLeaseQueueLeasePreferencePurgatoryError
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/d40a51d7f46724c09a296ebc19471f04/logTestLeaseQueueLeasePreferencePurgatoryError1624774253
    test_log_scope.go:76: use -show-logs to present logs inline
    lease_queue_test.go:176: condition failed to evaluate within 45s: from lease_queue_test.go:120: expected 40 leases on node 2, found 39
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d40a51d7f46724c09a296ebc19471f04/logTestLeaseQueueLeasePreferencePurgatoryError1624774253
--- FAIL: TestLeaseQueueLeasePreferencePurgatoryError (49.42s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseQueueLeasePreferencePurgatoryError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-44153

--- a/pkg/kv/kvserver/helpers_test.go
+++ b/pkg/kv/kvserver/helpers_test.go
@@ -160,10 +160,14 @@ func (s *Store) SplitQueuePurgatoryLength() int {
 	return s.splitQueue.PurgatoryLength()
 }
 
-// LeaseQueuePurgatoryLength returns the number of replicas in lease queue
-// purgatory.
-func (s *Store) LeaseQueuePurgatoryLength() int {
-	return s.leaseQueue.PurgatoryLength()
+// LeaseQueuePurgatory returns a map of RangeIDs representing the purgatory.
+func (s *Store) LeaseQueuePurgatory() map[roachpb.RangeID]struct{} {
+	defer s.leaseQueue.baseQueue.lockProcessing()()
+	m := make(map[roachpb.RangeID]struct{}, len(s.leaseQueue.baseQueue.mu.purgatory))
+	for k := range s.leaseQueue.baseQueue.mu.purgatory {
+		m[k] = struct{}{}
+	}
+	return m
 }
 
 // SetRaftLogQueueActive enables or disables the raft log queue.

--- a/pkg/kv/kvserver/lease_queue_test.go
+++ b/pkg/kv/kvserver/lease_queue_test.go
@@ -8,6 +8,7 @@ package kvserver_test
 import (
 	"context"
 	"fmt"
+	"sort"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -160,10 +161,32 @@ func TestLeaseQueueLeasePreferencePurgatoryError(t *testing.T) {
 	store := tc.GetFirstStoreFromServer(t, 0)
 	blockTransferTarget.Store(true)
 	setLeasePreferences(nextPreferredNode)
+	rangeIDs := func() map[roachpb.RangeID]struct{} {
+		rows := tdb.Query(t, `SELECT range_id FROM [ SHOW RANGES FROM TABLE t ]`)
+		var rangeID roachpb.RangeID
+		m := map[roachpb.RangeID]struct{}{}
+		for rows.Next() {
+			require.NoError(t, rows.Scan(&rangeID))
+			m[rangeID] = struct{}{}
+		}
+		require.NoError(t, rows.Err())
+		return m
+	}()
+	require.Len(t, rangeIDs, numRanges)
 	testutils.SucceedsSoon(t, func() error {
 		require.NoError(t, store.ForceLeaseQueueProcess())
-		if purgLen := store.LeaseQueuePurgatoryLength(); purgLen != numRanges {
-			return errors.Errorf("expected %d in purgatory but got %v", numRanges, purgLen)
+		purg := store.LeaseQueuePurgatory()
+		var missing []roachpb.RangeID
+		for k := range rangeIDs {
+			if _, ok := purg[k]; !ok {
+				missing = append(missing, k)
+			}
+		}
+		sort.Slice(missing, func(i, j int) bool {
+			return missing[i] < missing[j]
+		})
+		if len(missing) > 0 {
+			return errors.Errorf("replicas missing from purgatory: %v", missing)
 		}
 		return nil
 	})

